flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/SourceOperator.java

```diff
@@ -370,7 +411,17 @@ public DataInputStatus emitNext(DataOutput<OUT> output) throws Exception {
     private DataInputStatus emitNextNotReading(DataOutput<OUT> output) throws Exception {
         switch (operatingMode) {
             case OUTPUT_NOT_INITIALIZED:
-                currentMainOutput = eventTimeLogic.createMainOutput(output);
+                if (watermarkAlignmentParams.isEnabled()) {
+                    // Only wrap the output when watermark alignment is enabled, as otherwise this
+                    // introduces a small performance regression (probably because of an extra
+                    // virtual call)
+                    processingTimeService.scheduleWithFixedDelay(
+                            this::emitLatestWatermark,
+                            watermarkAlignmentParams.getUpdateInterval(),
+                            watermarkAlignmentParams.getUpdateInterval());
+                }
+                currentMainOutput =
+                        eventTimeLogic.createMainOutput(output, this::onWatermarkEmitted);
```

Comment:
Hard to make a call. It would be more descriptive to pass an empty callback, but it would be more complicated and `lastEmittedWatermark` would be not as consistent. As i can not decide which is better I would prefer leave it as it is.